cairo-libs/packages/cometbft/src/types.cairo

```diff
@@ -472,3 +526,348 @@ impl ValidatorSetAsProtoName of ProtoName<ValidatorSet> {
+                let vote = self.votes[i];
+
+                if *(vote.verified) {
+                    return false;
+                }
+
+                let signed_vote = vote.signed_vote;
+
+                let signature = signed_vote.signature;
+                let canonical_vote = signed_vote.vote;
+
+                let signed_bytes = ProtoCodecImpl::encode_with_length(canonical_vote);
+
+                let signed_array_u8 = byte_array_to_array_u8(@signed_bytes);
+
+                validator.verify_signature(signed_array_u8.span(), signature.span());
+
+                // TODO: set verified field to true
+
+                true
+            },
+        }
+    }
+}
+
+#[derive(Drop, Debug, Clone, PartialEq)]
+pub struct NonAbsentCommitVote {
+    signed_vote: SignedVote,
+    /// Flag indicating whether the signature has already been verified.
+    verified: bool,
+}
+
+#[generate_trait]
+pub impl NonAbsentCommitVoteImpl of NonAbsentCommitVoteTrait {
+    fn validator_id(self: @NonAbsentCommitVote) -> @AccountId {
+        self.signed_vote.validator_address
+    }
+}
+
+#[derive(Drop, Debug, Clone, PartialEq)]
+pub struct SignedVote {
+    vote: CanonicalVote,
+    validator_address: AccountId,
+    signature: Array<u8> // TODO: whether to define a Signature type?
+}
+
+#[derive(Drop, Debug, Clone, PartialEq, Default)]
+pub struct CanonicalVote {
+    /// Type of vote (prevote or precommit)
+    pub vote_type: VoteType,
+    /// Block height
+    pub height: i64,
+    /// Round
+    pub round: i64,
+    pub block_id: BlockId,
+    /// Timestamp
+    pub timestamp: Timestamp,
+    /// Chain ID
+    pub chain_id: ByteArray,
+}
+
+/// Type of votes
+#[derive(Drop, Debug, Clone, PartialEq, Default)]
+pub enum VoteType {
+    #[default]
+    Unknown,
+    /// Votes for blocks which validators observe are valid for a given round
+    Prevote,
+    /// Votes to commit to a particular block for a given round
+    Precommit,
+    Proposal,
+}
+
+pub impl VoteTypeIntoU32 of Into<@VoteType, u32> {
+    fn into(self: @VoteType) -> u32 {
+        match self {
+            VoteType::Unknown => 0,
+            VoteType::Prevote => 1,
+            VoteType::Precommit => 2,
+            VoteType::Proposal => 32,
+        }
+    }
+}
+
+pub impl U32TryIntoVoteType of TryInto<u32, VoteType> {
+    fn try_into(self: u32) -> Option<VoteType> {
+        if self == 0 {
+            return Option::Some(VoteType::Unknown);
+        } else if self == 1 {
+            return Option::Some(VoteType::Prevote);
+        } else if self == 2 {
+            return Option::Some(VoteType::Precommit);
+        } else if self == 32 {
+            return Option::Some(VoteType::Proposal);
+        } else {
+            return Option::None;
+        }
+    }
+}
+
+#[derive(Drop, Debug, Clone)]
+pub struct UntrustedBlockState {
+    pub signed_header: SignedHeader,
+    pub validators: ValidatorSet,
+    // pub next_validators: Option<ValidatorSet>,
```

Comment:
this field is present in `tendermint-rs` but never used in IBC and always set to `None` in ibc-rs.

I removed this, as `Option<T>` implements `Copy`, not `Clone`. And `ValidatoSet` only impls `Clone`.